src/lib.rs

```diff
@@ -471,13 +472,58 @@ impl Wal {
     pub fn clear(&mut self) -> Result<()> {
         self.truncate(self.first_index())
     }
+
+    /// Copy all files to the given path directory. directory should exist and be empty
+    pub fn copy_to_path<P>(&self, path: P) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        if fs::read_dir(path.as_ref())?.next().is_some() {
+            return Err(Error::new(
+                ErrorKind::AlreadyExists,
+                format!("path {:?} not empty", path.as_ref()),
+            ));
+        };
+
+        let open_segment_file = self.open_segment.segment.path().file_name().unwrap();
+        let close_segment_files: HashMap<_, _> = self
+            .closed_segments
+            .iter()
+            .map(|segment| {
+                (
+                    segment.segment.path().file_name().unwrap(),
+                    &segment.segment,
+                )
+            })
+            .collect();
+
+        for entry in fs::read_dir(self.path())? {
+            let entry = entry?;
+            if !entry.metadata()?.is_file() {
+                continue;
+            }
+
+            // if file is locked by any Segment, call copy_to_path on it
+            let entry_file_name = entry.file_name();
+            let dst_path = path.as_ref().to_owned().join(entry_file_name.clone());
+            if entry_file_name == open_segment_file {
+                self.open_segment.segment.copy_to_path(&dst_path)?;
+            } else if let Some(segment) = close_segment_files.get(entry_file_name.as_os_str()) {
+                segment.copy_to_path(&dst_path)?;
+            } else {
+                // if file is not locked by any Segment, just copy it
+                fs::copy(&entry.path(), &dst_path)?;
+            }
```

Comment:
Can we implement it in a way that copying through mmap only happens on Windows?

Or is that already happening in this configuration? Or does non-Windows suffer from this issue as well?

Asking because only doing it on Windows would reduce complexity on Linux.

Reply:
The bug is on docker with windows host. I cannot propose effective way how to detect: is this linux under docker with windows host or not

Reply:
Oh yes, you've clarified this before. You're totally right. Thanks for elaborating once again!